### ISSUE TITLE
TextField Previous and Next not appear

### ISSUE THREAD

**Issue by liufengyu** [hackiftekhar/IQKeyboardManager#549]
on ViewController when I add TextField to a subView of self.view

**Comment by hackiftekhar** [comment]
Prev Next will appear if there are more than 1 sibling textField

**Comment by liufengyu** [comment]
I add 3 textField but not appear

**Comment by hackiftekhar** [comment]
You added 3 textField but doesn't appear? when you added to self.view they appeared? didn't understand. Both are looking like same statement.

**Comment by oliverfoggin** [comment]
Is there any way to get IQKeyboardManager to search up and down the view hierarchy? I have a view laid out with multiple nested UIStackViews (a common use case for stack views) and I can't find a way to get the next rev to appear.

Would be awesome to have a way of manually setting them like a delegate callback "nextTextFieldForCurrentTextField" etc...

**Comment by oliverfoggin** [comment]
I added a couple lines of code... not in a PR yet but possibly soon.

Anyway...

    ///------------------
    /// MARK: AutoToolbar
    ///------------------
    
    /**	Get all UITextField/UITextView siblings of textFieldView. */
    private func responderViews()-> [UIView]? {
        
        if let overriddenTextFields = overrideIteratingTextFields {
            return overriddenTextFields
        }
        ... rest of the method...

and a new property...

    public var overrideIteratingTextFields: [UIView]?

And all I need to do is...
        
        IQKeyboardManager.sharedManager().overrideIteratingTextFields = [
            nameField,
            cardNumberField,
            expiryDateField,
            cvvField
        ]

and it works fine. Just need to set it back to nil in `viewWillDisappear`

**Comment by hackiftekhar** [comment]
That's a bit of extra manual management and need to take extra care in viewWillDisappear. This is not we want in a library, @oliverfoggin We may need to come with a generic solution which should do the trick without any external textField references.

**Comment by oliverfoggin** [comment]
@hackiftekhar true. But in this case the automatic part of the library isn't working. You could make it recursively search for all UITextFields but not sure how that would perform (also a possibility of getting text fields that currently aren't visible etc...)

This was a quick hack for me and works until something better is done.

**Comment by hackiftekhar** [comment]
Yeah you totally correct *It's a quick hack* this is why it's not suitable for a open source library used by many developers. We may need to come up with a generic solution by combining toolbarManageBehaviour with IQPreviousNextView.

**Comment by oliverfoggin** [comment]
@hackiftekhar I didn't say it was suitable. I merely pointed out how I had forced it to work. I had 20 minutes to get this working. I wouldn't have submitted this but something similar. Anyway, it is working for me now and if I remove the update then it doesn't work so I'm happy. Do whatever you think is necessary.

**Comment by oliverfoggin** [comment]
@hackiftekhar please could you explain what the `toolbarPreviousNextAllowedClasses` property does?

I have added this... 

        IQKeyboardManager.sharedManager().toolbarPreviousNextAllowedClasses = [UIStackView.self, UIView.self]

But it is still not showing prev or next arrows.

Going to have to put my hack in again to fix this. Thanks

**Comment by hackiftekhar** [comment]
This property treats all it's inner subviews as part of siblings to navigate through. I'm not sure why it doesn't work for you.

**Comment by oliverfoggin** [comment]
@hackiftekhar I think the problem (possibly) is that it only goes one level deep.

My view hierarchy is...

```
UIStackView
    -UIStackView
        -UITextField
    -UIStackView
        -UITextField
```

So it would need to go up twice and down twice to find the next/prev text field.

It's actually more complex than this but it shows as an example.

**Comment by hackiftekhar** [comment]
ok, I think you should put all this hierarchy inside IQPreviousNextView (On top of the most top UIStackView), so it will find all inner textFields.

**Comment by oliverfoggin** [comment]
@hackiftekhar ah! That works! Thanks :D

**Comment by viral105** [comment]
What to do in objective c for same problem ?

**Comment by hackiftekhar** [comment]
Things are same for ObjC and Swift, so you can follow same approach in both languages.

**Comment by viral105** [comment]
Thanks for reply.
I have tried following in objective c.
[IQKeyboardManager sharedManager] = [tfd1,tfd2,tfd3,tfd4];

but it is giving me following error.
Expected Identifier.

**Comment by hackiftekhar** [comment]
@viral105 please check solution regarding `IQPreviousNextView`. You are currently doing wrong.

**Comment by viral105** [comment]
@hackiftekhar I am not using UIStackView but my textfields are subview of different UIView. So where should I found the code for that ?

**Comment by hackiftekhar** [comment]
You could check *CustomView Example* in demo or you can read documentation here:-
https://github.com/hackiftekhar/IQKeyboardManager/blob/master/MANUAL%20MANAGEMENT.md

**Comment by oliverfoggin** [comment]
@viral105 the solution (that was mentioned long ago) is to use the IQPreviousNextView and put all your text fields inside that.

This is a very easy solution and works.

I'm not even sure what your code... `[IQKeyboardManager sharedManager] = [tfd1,tfd2,tfd3,tfd4];` is trying to do? What do you think it is doing?

Just put the controls inside the `IQPreviousNextView`.

**Comment by viral105** [comment]
Its Done. @oliverfoggin comment helped me to resolve my problem.
Thanks @oliverfoggin  and @hackiftekhar for your time and help.

**Comment by kumardiwa** [comment]
I've 5 textFileds in a single StackView and previous and next button is not working. Although I've written **IQKeyboardManager.sharedManager().previousNextDisplayMode = .alwaysShow** but still not working.

**Comment by hackiftekhar** [comment]
Your all textField might not be in same superview I think?

**Comment by xeieshan** [comment]
overrideIteratingTextFields cannot work for me. Kindly help!

**Comment by karthisiva** [comment]
> IQPreviousNextView

Any sample for this?

**Comment by Bergasms** [comment]
I'm curious why the Author thinks embedding views in a IQPreviousNextView is not considered too much manual work but sticking references in an array is considered too much manual work. For a library presenting itself as automatic i think buggering around with classes in a storyboard is more invasive and prone to future problems than handling a manual override in your code. It's certainly easier to remove the code based condition and the compiler will catch problems for you at that level. Putting custom things in a storyboard is a great way to give someone a runtime crash in the future and you also cannot document it for anyone coming past making changes to the code.       

@oliverfoggin presents a really simple solution that will work for any sort of complicated layout where you have the references available, and in 99% of the cases you will have the references available because input fields without references are not super useful.     

Regardless, the IQPreviousNextView solution does work, but now I have had to pollute one of our libraries with another dependency just to add it into the storyboard, and write a note explaining the footgun that's now hidden in one of the libraries in an inobvious way that the compiler won't catch.
